core/chains/evm/gas/rollups/custom_calldata_da_oracle.go

```diff
@@ -0,0 +1,167 @@
+package rollups
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+	evmclient "github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+	evmconfig "github.com/smartcontractkit/chainlink/v2/core/chains/evm/config"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/config/toml"
+)
+
+type customCalldataDAOracle struct {
+	services.StateMachine
+	client     l1OracleClient
+	pollPeriod time.Duration
+	logger     logger.SugaredLogger
+
+	daOracleConfig evmconfig.DAOracle
+	daGasPriceMu   sync.RWMutex
+	daGasPrice     priceEntry
+
+	chInitialized chan struct{}
+	chStop        services.StopChan
+	chDone        chan struct{}
+}
+
+// NewCustomCalldataDAOracle creates a new custom calldata DA oracle. The CustomCalldataDAOracle fetches gas price from
+// whatever function is specified in the DAOracle's CustomGasPriceCalldata field. This allows for more flexibility when
+// chains have custom DA gas calculation methods.
+func NewCustomCalldataDAOracle(lggr logger.Logger, ethClient l1OracleClient, daOracleConfig evmconfig.DAOracle) (*customCalldataDAOracle, error) {
+	if daOracleConfig.OracleType() != toml.DAOracleCustomCalldata {
+		return nil, fmt.Errorf("expected CustomCalldata oracle type, got %s", daOracleConfig.OracleType())
+	}
+	if daOracleConfig.CustomGasPriceCalldata() == "" {
+		return nil, fmt.Errorf("custom gas price calldata is required")
```

Comment:
nit: might be better to refer to the config name here so the user can tell what needs to be set more explicitly
```suggestion
		return nil, fmt.Errorf("CustomGasPriceCalldata is required")
```